copystructure.go

```diff
@@ -48,14 +49,25 @@ type walker struct {
 	vals        []reflect.Value
 	cs          []reflect.Value
 	ps          []bool
+	locks       []sync.Locker
 }
 
 func (w *walker) Enter(l reflectwalk.Location) error {
 	w.depth++
+	// this normally appends only once, but make sure we start with enough
+	// space to use depth as an index
+	for w.depth >= len(w.locks) {
+		w.locks = append(w.locks, nil)
+	}
```

Comment:
When I actually coded it out your solution is actually a lot shorter, so its probably fine at the expense of more CPU cycles probably but we don't need to optimize for that here.